File: BackEnd-Siukat/services/ukt.go
package services

import (
	"BackEnd-Siukat/config"
	"BackEnd-Siukat/models"
)

type UKTService struct{}

func vFirst(gajiIbu, gajiAyah, sampinganIbu, sampinganAyah, uangWali, gajiSendiri, tanggungan float64) float64 {
	// Formula: (gaji_ibu + gaji_ayah + sampingan_ibu + sampingan_ayah + uang_wali + gaji_sendiri) / tanggungan
	// Safety net for Golang to prevent panic div by zero
	if tanggungan == 0 {
		tanggungan = 1
	}
	return (gajiIbu + gajiAyah + sampinganIbu + sampinganAyah + uangWali + gajiSendiri) / tanggungan
}

func vSecond(pbb, kk, kontrak, bobotPbb, bobotKontrak float64) float64 {
	// Formula: bobot_pbb * ((pbb / 12) * kk) + bobot_kontrak * (kontrak / 12)
	return bobotPbb*((pbb/12)*kk) + bobotKontrak*(kontrak/12)
}

func vThird(listrik float64) float64 {
	// Formula: listrik / 3
	return listrik / 3
}

func vFourthFifth(kendaraan float64) float64 {
	// Formula: kendaraan / 12
	return kendaraan / 12
}

func decisionMaker(ikb float64, ukt models.Ukt) string {
	uktTemp := "VIII"
	if ikb <= float64(ukt.I) {
		uktTemp = "I"
	} else if ikb <= float64(ukt.II) {
		uktTemp = "II"
	} else if ikb <= float64(ukt.III) {
		uktTemp = "III"
	} else if ikb <= float64(ukt.IV) {
		uktTemp = "IV"
	} else if ikb <= float64(ukt.V) {
		uktTemp = "V"
	} else if ikb <= float64(ukt.VI) {
		uktTemp = "VI"
	} else if ikb <= float64(ukt.VII) {
		uktTemp = "VII"
	} else if ikb <= float64(ukt.VIII) {
		uktTemp = "VIII"
	}
	return uktTemp
}

// ComputeUkt adalah translasi 1-to-1 dari versi Node.js (ukt.prototype.computeUkt)
func (s *UKTService) ComputeUkt(noPeserta string, atribut string) (map[string]interface{}, error) {
	db := config.DB
	data := make(map[string]interface{})

	var cmahasiswa models.CMahasiswa
	if err := db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&cmahasiswa).Error; err != nil {
		return nil, err
	}
	data["cmahasiswa"] = cmahasiswa

	var ayah models.Ayah
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&ayah)
	data["ayah"] = ayah

	var ibu models.Ibu
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&ibu)
	data["ibu"] = ibu

	var wali models.Wali
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&wali)
	data["wali"] = wali

	var rumah models.Rumah
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&rumah)
	data["rumah"] = rumah

	var listrik models.Listrik
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&listrik)
	data["listrik"] = listrik

	var kendaraan models.Kendaraan
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&kendaraan)
	data["kendaraan"] = kendaraan

	var pendukung models.Pendukung
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&pendukung)
	data["pendukung"] = pendukung

	var bobot models.Bobot
	db.Where("id = ?", 1).First(&bobot) // Default fetch id=1
	data["bobot"] = bobot

	var ukt models.Ukt
	db.Where("major_id = ? AND entrance = ?", cmahasiswa.ProdiCmahasiswa, cmahasiswa.JalurCmahasiswa).First(&ukt)
	data["ukt"] = ukt

	// Kalkulasi V1 - V5
	v1 := vFirst(float64(ibu.PenghasilanIbu), float64(ayah.PenghasilanAyah), float64(ibu.SampinganIbu), float64(ayah.SampinganAyah), float64(wali.KesanggupanWali), float64(cmahasiswa.PenghasilanCmahasiswa), float64(pendukung.Tanggungan))
	v2 := vSecond(float64(rumah.BiayaPbb), float64(rumah.JumlahKepalaKeluarga), float64(rumah.BiayaKontrak), float64(bobot.Pbb), float64(bobot.Kontrak))
	v3 := vThird(float64(listrik.Pengeluaran))
	v4 := vFourthFifth(float64(kendaraan.PajakMotor))
	v5 := vFourthFifth(float64(kendaraan.PajakMobil))

	// Bobot Av1 - Ev5
	av1 := v1 * 6 * bobot.A
	bv2 := v2 * 6 * bobot.B
	cv3 := v3 * 6 * bobot.C
	dv4 := v4 * 6 * bobot.D
	ev5 := v5 * 6 * bobot.E

	// Hitung IKB
	ikb := av1 + bv2 + cv3 + dv4 + ev5

	// Decision Golongan
	choosenUkt := decisionMaker(ikb, ukt)

	// Save calculation metrics to data map
	data["v1"] = v1
	data["v2"] = v2
	data["v3"] = v3
	data["v4"] = v4
	data["v5"] = v5
	data["av1"] = av1
	data["bv2"] = bv2
	data["cv3"] = cv3
	data["dv4"] = dv4
	data["ev5"] = ev5
	data["ikb"] = ikb
	data["choosenUkt"] = choosenUkt

	// Translating Value.update(...)
	valueUpdate := map[string]interface{}{
		"v1": v1, "v2": v2, "v3": v3, "v4": v4, "v5": v5,
		"av1": av1, "bv2": bv2, "cv3": cv3, "dv4": dv4, "ev5": ev5,
		"ikb": ikb,
	}
	var valueResult models.Value
	db.Model(&models.Value{}).Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).Updates(valueUpdate)
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&valueResult)
	data["value"] = valueResult

	// Translating CMahasiswa.update(golongan_id)
	db.Model(&models.CMahasiswa{}).Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).Update("golongan_id", choosenUkt)
	
	var updatedCMhs models.CMahasiswa
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&updatedCMhs)
	data["uktCmahasiswa"] = updatedCMhs

	return data, nil
}

// JustCompute is exactly the same without the DB Updates
func (s *UKTService) JustCompute(noPeserta string, atribut string) (map[string]interface{}, error) {
	// ... (Sama seperti ComputeUkt di awal)
	db := config.DB
	data := make(map[string]interface{})

	var cmahasiswa models.CMahasiswa
	if err := db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&cmahasiswa).Error; err != nil {
		return nil, err
	}
	var ayah models.Ayah
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&ayah)
	var ibu models.Ibu
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&ibu)
	var wali models.Wali
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&wali)
	var rumah models.Rumah
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&rumah)
	var listrik models.Listrik
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&listrik)
	var kendaraan models.Kendaraan
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&kendaraan)
	var pendukung models.Pendukung
	db.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&pendukung)
	var bobot models.Bobot
	db.Where("id = ?", 1).First(&bobot)
	
	var ukt models.Ukt
	db.Where("major_id = ? AND entrance = ?", cmahasiswa.ProdiCmahasiswa, cmahasiswa.JalurCmahasiswa).First(&ukt)

	v1 := vFirst(float64(ibu.PenghasilanIbu), float64(ayah.PenghasilanAyah), float64(ibu.SampinganIbu), float64(ayah.SampinganAyah), float64(wali.KesanggupanWali), float64(cmahasiswa.PenghasilanCmahasiswa), float64(pendukung.Tanggungan))
	v2 := vSecond(float64(rumah.BiayaPbb), float64(rumah.JumlahKepalaKeluarga), float64(rumah.BiayaKontrak), float64(bobot.Pbb), float64(bobot.Kontrak))
	v3 := vThird(float64(listrik.Pengeluaran))
	v4 := vFourthFifth(float64(kendaraan.PajakMotor))
	v5 := vFourthFifth(float64(kendaraan.PajakMobil))

	av1 := v1 * 6 * bobot.A
	bv2 := v2 * 6 * bobot.B
	cv3 := v3 * 6 * bobot.C
	dv4 := v4 * 6 * bobot.D
	ev5 := v5 * 6 * bobot.E

	ikb := av1 + bv2 + cv3 + dv4 + ev5
	choosenUkt := decisionMaker(ikb, ukt)

	// Save data
	data["cmahasiswa"], data["ayah"], data["ibu"], data["wali"] = cmahasiswa, ayah, ibu, wali
	data["rumah"], data["listrik"], data["kendaraan"], data["pendukung"] = rumah, listrik, kendaraan, pendukung
	data["bobot"], data["ukt"] = bobot, ukt
	data["v1"], data["v2"], data["v3"], data["v4"], data["v5"] = v1, v2, v3, v4, v5
	data["av1"], data["bv2"], data["cv3"], data["dv4"], data["ev5"] = av1, bv2, cv3, dv4, ev5
	data["ikb"], data["choosenUkt"] = ikb, choosenUkt

	return data, nil
}

// ByCmahasiswa resolves the UKT nominal for a single student profile
func (s *UKTService) ByCmahasiswa(mhs models.CMahasiswa) (int, error) {
	var ukt models.Ukt
	err := config.DB.Where("major_id = ? AND entrance = ?", mhs.ProdiCmahasiswa, mhs.JalurCmahasiswa).First(&ukt).Error
	if err != nil {
		return 0, err
	}

	switch mhs.GolonganID {
	case "I":
		return ukt.I, nil
	case "II":
		return ukt.II, nil
	case "III":
		return ukt.III, nil
	case "IV":
		return ukt.IV, nil
	case "V":
		return ukt.V, nil
	case "VI":
		return ukt.VI, nil
	case "VII":
		return ukt.VII, nil
	case "VIII":
		return ukt.VIII, nil
	default:
		return 0, nil
	}
}

// PerCmahasiswa handles batch resolution of UKT nominals for a slice of students
func (s *UKTService) PerCmahasiswa(students []models.CMahasiswa) ([]map[string]interface{}, error) {
	var results []map[string]interface{}
	for _, mhs := range students {
		nominal, _ := s.ByCmahasiswa(mhs)
		item := map[string]interface{}{
			"student": mhs,
			"nominal": nominal,
		}
		results = append(results, item)
	}
	return results, nil
}

File: ZhammDataBeLama/siukat-service-dev_snmptn_2022/services/ukt.js
var Ukt = require("../models/ukt");
var CMahasiswa = require("../models/cmahasiswa");
var Ayah = require("../models/ayah");
var Ibu = require("../models/ibu");
var Wali = require("../models/wali");
var Rumah = require("../models/rumah");
var Listrik = require("../models/listrik");
var Kendaraan = require("../models/kendaraan");
var Pendukung = require("../models/pendukung");
var Bobot = require("../models/bobot");
var Value = require("../models/value");

var ukt = function () {};

const vFirst = (
  gaji_ibu,
  gaji_ayah,
  sampingan_ibu,
  sampingan_ayah,
  uang_wali,
  gaji_sendiri,
  tanggungan
) => {
  return (
    (gaji_ibu +
      gaji_ayah +
      sampingan_ibu +
      sampingan_ayah +
      uang_wali +
      gaji_sendiri) /
    tanggungan
  );
};

const vSecond = (pbb, kk, kontrak, bobot_pbb, bobot_kontrak) => {
  return bobot_pbb * ((pbb / 12) * kk) + bobot_kontrak * (kontrak / 12);
};

const vThird = (listrik) => {
  return listrik / 3;
};

const vFourthFifth = (kendaraan) => {
  return kendaraan / 12;
};

/*const decisionMaker = (ikb, ukt) => {
    var uktCategories = {}
    uktCategories.I = ukt.I
    uktCategories.II = ukt.II
    uktCategories.III = ukt.III
    uktCategories.IV = ukt.IV
    uktCategories.V = ukt.V
    uktCategories.VI = ukt.VI
    uktCategories.VII = ukt.VII
    uktCategories.VIII = ukt.VIII

    // var selisih = 99999999
    var ukt_temp = 0
    if(ikb <= uktCategories.I){
        ukt_temp = 'I'            
    }else if(ikb <= 1500000){
        ukt_temp = 'II'
    }else if(ikb <= uktCategories.III){
        ukt_temp = 'III'
    }else if(ikb <= uktCategories.IV){
        ukt_temp = 'IV'
    }else if(ikb <= uktCategories.V){
        ukt_temp = 'V'
    }else if(ikb <= uktCategories.VI){
        ukt_temp = 'VI'
    }else if(ikb <= uktCategories.VII){
        ukt_temp = 'VII'
    }else if(ikb <= uktCategories.VIII){
        ukt_temp = 'VIII'
    }else{
        ukt_temp = 'VIII'
    }
    // for(var key in uktCategories){
    //     if(key === "II"){
    //         temp = Math.abs(1500000 - ikb)
    //     }else{
    //         temp = Math.abs(uktCategories[key] - ikb)
    //     }

    //     if(temp <= selisih){
    //         selisih = temp
    //         ukt_temp = key
    //     }
    // }

    return ukt_temp
}*/

// const decisionMaker = (ikb, ukt) => {
//     var uktCategories = {}
//     uktCategories.I = ukt.I
//     uktCategories.II = ukt.II
//     uktCategories.III = ukt.III
//     uktCategories.IV = ukt.IV
//     uktCategories.V = ukt.V
//     uktCategories.VI = ukt.VI
//     uktCategories.VII = ukt.VII
//     uktCategories.VIII = ukt.VIII

//     // var selisih = 99999999
//     var ukt_temp = 0
//     if (ikb <= uktCategories.I) {
//         ukt_temp = 'I'
//     } else if (ikb <= 1500000) {
//         ukt_temp = 'II'
//     } else if (ikb <= uktCategories.III) {
//         ukt_temp = 'III'
//     } else if (ikb <= uktCategories.IV) {
//         ukt_temp = 'IV'
//     } else if (ikb <= uktCategories.V) {
//         ukt_temp = 'V'
//     } else if (ikb <= uktCategories.VI) {
//         ukt_temp = 'VI'
//     } else if (ikb <= uktCategories.VII) {
//         ukt_temp = 'VII'
//     } else if (ikb <= uktCategories.VIII) {
//         ukt_temp = 'VIII'
//     } else {
//         ukt_temp = 'VIII'
//     }
// for(var key in uktCategories){
//     if(key === "II"){
//         temp = Math.abs(1500000 - ikb)
//     }else{
//         temp = Math.abs(uktCategories[key] - ikb)
//     }

//     if(temp <= selisih){
//         selisih = temp
//         ukt_temp = key
//     }
// }

//     return ukt_temp
// }

const decisionMaker = (ikb, ukt) => {
  var uktCategories = {};
  uktCategories.I = ukt.I;
  uktCategories.II = ukt.II;
  uktCategories.III = ukt.III;
  uktCategories.IV = ukt.IV;
  uktCategories.V = ukt.V;
  uktCategories.VI = ukt.VI;
  uktCategories.VII = ukt.VII; //null
  uktCategories.VIII = ukt.VIII; //null

  var ukt_temp = 0;
  var category_temp = Object.entries(uktCategories);

  // for(var i = 0; i <category_temp.length; i++){
  //     if(category_temp[i][1]){
  //         console.log('woy' + i)
  //         if(category_temp[i][0] === 'II' && ikb <= 1500000){
  //             ukt_temp = category_temp[i][0]
  //             break
  //         }else if(ikb <= category_temp[i][1]){
  //             ukt_temp = 'II'
  //             break
  //         }
  //     }else{
  //         ukt_temp = category_temp[i-1][0]
  //         break
  //     }
  // }

  // var selisih = 99999999
  ukt_temp = 0;
  if (ikb <= uktCategories.I) {
    ukt_temp = "I";
  } else if (ikb <= uktCategories.II) {
    ukt_temp = "II";
  } else if (ikb <= uktCategories.III) {
    ukt_temp = "III";
  } else if (ikb <= uktCategories.IV) {
    ukt_temp = "IV";
  } else if (ikb <= uktCategories.V) {
    ukt_temp = "V";
  } else if (ikb <= uktCategories.VI) {
    ukt_temp = "VI";
  } else if (ikb <= uktCategories.VII) {
    ukt_temp = "VII";
  } else if (ikb <= uktCategories.VIII) {
    ukt_temp = "VIII";
  } else {
    ukt_temp = "VIII";
  }
  // for(var key in uktCategories){
  //     if(key === "II"){
  //         temp = Math.abs(1500000 - ikb)
  //     }else{
  //         temp = Math.abs(uktCategories[key] - ikb)
  //     }

  //     if(temp <= selisih){
  //         selisih = temp
  //         ukt_temp = key
  //     }
  // }

  return ukt_temp;
};

ukt.prototype.computeUkt = (no_peserta, atribut) => {
  return new Promise((resolve, reject) => {
    var data = {};
    CMahasiswa.findOne({ where: { no_peserta: no_peserta, atribut: atribut } })
      .then((mhs) => {
        data.cmahasiswa = mhs;
        // console.log(mhs);
        return Ayah.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.ayah = response;
        return Ibu.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.ibu = response;
        return Wali.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.wali = response;
        return Rumah.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.rumah = response;
        return Listrik.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.listrik = response;
        return Kendaraan.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.kendaraan = response;
        return Pendukung.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.pendukung = response;
        return Bobot.findOne({ where: { id: 1 } });
      })
      .then((response) => {
        data.bobot = response;
        console.log(no_peserta);
        return Ukt.findOne({
          where: {
            major_id: data.cmahasiswa.prodi_cmahasiswa,
            entrance: data.cmahasiswa.jalur_cmahasiswa,
          },
        });
      })
      .then((response) => {
        data.ukt = response;

        data.v1 = vFirst(
          data.ibu.penghasilan_ibu,
          data.ayah.penghasilan_ayah,
          data.ibu.sampingan_ibu,
          data.ayah.sampingan_ayah,
          data.wali.kesanggupan_wali,
          data.cmahasiswa.penghasilan_cmahasiswa,
          data.pendukung.tanggungan
        );
        data.v2 = vSecond(
          data.rumah.biaya_pbb,
          data.rumah.jumlah_kepala_keluarga,
          data.rumah.biaya_kontrak,
          data.bobot.pbb,
          data.bobot.kontrak
        );
        data.v3 = vThird(data.listrik.pengeluaran);
        data.v4 = vFourthFifth(data.kendaraan.pajak_motor);
        data.v5 = vFourthFifth(data.kendaraan.pajak_mobil);

        data.av1 = data.v1 * 6 * data.bobot.a;
        data.bv2 = data.v2 * 6 * data.bobot.b;
        data.cv3 = data.v3 * 6 * data.bobot.c;
        data.dv4 = data.v4 * 6 * data.bobot.d;
        data.ev5 = data.v5 * 6 * data.bobot.e;

        data.ikb = data.av1 + data.bv2 + data.cv3 + data.dv4 + data.ev5;

        data.choosenUkt = decisionMaker(data.ikb, data.ukt);

        return Value.update(
          {
            v1: data.v1,
            v2: data.v2,
            v3: data.v3,
            v4: data.v4,
            v5: data.v5,
            av1: data.av1,
            bv2: data.bv2,
            cv3: data.cv3,
            dv4: data.dv4,
            ev5: data.ev5,
            ikb: data.ikb,
          },
          { where: { no_peserta: no_peserta, atribut: atribut } }
        );
      })
      .then((response) => {
        data.value = response;

        return CMahasiswa.update(
          {
            golongan_id: data.choosenUkt,
          },
          { where: { no_peserta: no_peserta, atribut: atribut } }
        );
      })
      .then((response) => {
        data.uktCmahasiswa = response;
        resolve(data);
      })
      .catch((err) => {
        reject(err);
      });
  });
};

ukt.prototype.justCompute = (no_peserta, atribut) => {
  return new Promise((resolve, reject) => {
    var data = {};
    CMahasiswa.findOne({ where: { no_peserta: no_peserta, atribut: atribut } })
      .then((response) => {
        data.cmahasiswa = response;
        return Ayah.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.ayah = response;
        return Ibu.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.ibu = response;
        return Wali.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.wali = response;
        return Rumah.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.rumah = response;
        return Listrik.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.listrik = response;
        return Kendaraan.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.kendaraan = response;
        return Pendukung.findOne({
          where: { no_peserta: no_peserta, atribut: atribut },
        });
      })
      .then((response) => {
        data.pendukung = response;
        return Bobot.findOne({ where: { id: 1 } });
      })
      .then((response) => {
        data.bobot = response;
        return Ukt.findOne({
          where: {
            major_id: data.cmahasiswa.prodi_cmahasiswa,
            entrance: data.cmahasiswa.jalur_cmahasiswa,
          },
        });
      })
      .then((response) => {
        data.ukt = response;

        data.v1 = vFirst(
          data.ibu.penghasilan_ibu,
          data.ayah.penghasilan_ayah,
          data.ibu.sampingan_ibu,
          data.ayah.sampingan_ayah,
          data.wali.kesanggupan_wali,
          data.cmahasiswa.penghasilan_cmahasiswa,
          data.pendukung.tanggungan
        );
        data.v2 = vSecond(
          data.rumah.biaya_pbb,
          data.rumah.jumlah_kepala_keluarga,
          data.rumah.biaya_kontrak,
          data.bobot.pbb,
          data.bobot.kontrak
        );
        data.v3 = vThird(data.listrik.pengeluaran);
        data.v4 = vFourthFifth(data.kendaraan.pajak_motor);
        data.v5 = vFourthFifth(data.kendaraan.pajak_mobil);

        data.av1 = data.v1 * 6 * data.bobot.a;
        data.bv2 = data.v2 * 6 * data.bobot.b;
        data.cv3 = data.v3 * 6 * data.bobot.c;
        data.dv4 = data.v4 * 6 * data.bobot.d;
        data.ev5 = data.v5 * 6 * data.bobot.e;

        data.ikb = data.av1 + data.bv2 + data.cv3 + data.dv4 + data.ev5;

        data.choosenUkt = decisionMaker(data.ikb, data.ukt);
      })
      .then((response) => {
        resolve(data);
      })
      .catch((err) => {
        reject(err);
      });
  });
};

ukt.prototype.byCmahasiswa = (cmahasiswa) => {
  let columnUKT = cmahasiswa.golongan_id
    ? [cmahasiswa.golongan_id, "nominal"]
    : "*";
  Ukt.findOne({
    where: {
      major_id: cmahasiswa.prodi_cmahasiswa,
      entrance: cmahasiswa.jalur_cmahasiswa,
    },
    attributes: [columnUKT],
  })
    .then((result) => {
      resolve(result);
    })
    .catch((err) => {
      reject(err);
    });
};

ukt.prototype.perCmahasiswa = (response, convertToJSON = true) => {
  let rows = [],
    promises = [];
  for (let i in response.rows) {
    let columnUKT = response.rows[i].golongan_id
      ? [response.rows[i].golongan_id, "nominal"]
      : "*";
    var temp = new Promise((resolve, reject) => {
      Ukt.findOne({
        where: {
          major_id: response.rows[i].prodi_cmahasiswa,
          entrance: response.rows[i].jalur_cmahasiswa,
        },
        attributes: [columnUKT],
      }).then((result) => {
        if (convertToJSON) response.rows[i] = response.rows[i].toJSON();
        rows.push(response.rows[i]);
        resolve(result);
      });
    });
    promises.push(temp);
  }

  return {
    rows: rows,
    promises: promises,
  };
};

ukt.prototype.rupiah = (angka) => {
  if (angka !== undefined) {
    var value = angka.toString().replace(".", "");
    var rupiah = "";
    var angkarev = value.toString().split("").reverse().join("");
    for (var i = 0; i < angkarev.length; i++)
      if (i % 3 === 0) rupiah += angkarev.substr(i, 3) + ".";
    return (
      "Rp " +
      rupiah
        .split("", rupiah.length - 1)
        .reverse()
        .join("")
    );
  }
};

module.exports = ukt;
